src/strings/nfa-strings.cc

```diff
@@ -145,3 +145,49 @@ void ShortestWordsMap::update_current_words(LengthWordsPair& act, const LengthWo
     }
     act.first = dst.first + 1;
 }
+
+
+std::set<std::pair<int, int>> Mata::Strings::get_word_lengths(const Nfa::Nfa& aut) {
+    Nfa::Nfa one_letter;
+    aut.get_one_letter_aut(one_letter);
+    one_letter = determinize(one_letter);
+    one_letter.trim();
+
+    std::set<std::pair<int, int>> ret;
+    std::vector<int> handles(one_letter.size(), 0); // initialized to 0
+    std::deque<Nfa::State> worklist(one_letter.initial.begin(), one_letter.initial.end()); 
+    std::set<Nfa::State> visited;
+    int cnt = 0; // handle counter
+    int loop_size = 0; // loop size
+    int loop_start = -1; // cnt where the loop starts
+
+    while(!worklist.empty()) {
+        Nfa::State state = worklist.front();
+        worklist.pop_front();
+        visited.insert(state);
+
+        handles[state] = cnt++;
+        Nfa::Post post = one_letter.delta[state];
+        assert(post.size() <= 1);
+        for(const Move& move : post) {
+            assert(move.targets.size() == 1);
+            Nfa::State tgt = *move.targets.begin();
+
+            if(visited.find(tgt) == visited.end()) {
+                worklist.push_back(tgt);
+            } else {
+                loop_start = handles[tgt];
+                loop_size = cnt - handles[tgt];
```

Comment:
Is there an implicit assumption that the final states are processed at last, so the `loop_*` variables can be used in the next `for` cycle to dermine loops from the final states?

Reply:
I am not totally sure what you mean. After traversing the whole automaton (while loop), `loop_start` and `loop_size` should be properly set (they are unique values for the whole automaton)